pykafka/producer.py

```diff
@@ -111,15 +158,112 @@ def __repr__(self):
+                        self, partition.leader)
+            self._running = True
+
+    @raise_worker_exceptions
+    def stop(self, wait=True):
+        """Mark the producer as stopped
+
+        :param wait: Whether to wait for all inflight messages to be sent
+            before returning
+        :type wait: bool
+        """
+        self._running = False
+        if wait:
+            self._wait_all()
+
+    @raise_worker_exceptions
+    def produce(self, messages):
+        """Produce a set of messages.
+
+        :param messages: The messages to produce
+        :type messages: Iterable of str or (str, str) tuples
+        """
+        def _partition_messages():
+            """Assign messages to partitions using the partitioner."""
+            partitions = self._topic.partitions.values()
+            for message in messages:
+                if isinstance(message, basestring):
+                    key = None
+                    value = message
+                else:
+                    key, value = message
+                yield (key, str(value)), self._partitioner(partitions, message).id
+        if not self._running:
+            raise ProducerStoppedException()
+        pairs = itertools.izip(self._produce(_partition_messages()), messages)
+        for message_enqueued in pairs:
+            yield message_enqueued
+
+    @raise_worker_exceptions
+    def _produce(self, message_partition_tups):
+        """Enqueue a set of messages for the relevant brokers
+
+        :param message_partition_tups: Messages with partitions assigned.
+        :type message_partition_tups: Iterable of ((str, str), int) tuples
+        """
+        for kv, partition_id in message_partition_tups:
+            leader_id = self._topic.partitions[partition_id].leader.id
+            yield self._owned_brokers[leader_id].enqueue([(kv, partition_id)])
+
+    @raise_worker_exceptions
+    def _prepare_request(self, message_batch, owned_broker):
+        """Prepare a request and send it to the broker
+
+        :param message_batch: An iterable of messages to send
+        :type message_batch: iterable of `((key, value), partition_id)` tuples
+        :param owned_broker: The `OwnedBroker` to which to send the request
+        :type owned_broker: :class:`pykafka.producer.OwnedBroker`
+        """
+        request = ProduceRequest(
+            compression_type=self._compression,
+            required_acks=self._required_acks,
+            timeout=self._ack_timeout_ms
+        )
+        for (key, value), partition_id in message_batch:
+            request.add_message(
+                Message(value, partition_key=key),
+                self._topic.name,
+                partition_id
+            )
+        log.debug("Sending %d messages to broker %d",
+                  len(message_batch), owned_broker.broker.id)
+        self._send_request(request, 0, owned_broker)
+
+    @raise_worker_exceptions
```

Comment:
Yes, and I've come to believe that my reasoning for doing that was flawed. I'll just keep this decorator on public methods.